File: picat/benchmark/code/reverse.hs
import Text.Printf
import Control.Exception
import System.CPUTime

normal_reverse []    = []
normal_reverse (x:t) = (normal_reverse t) ++ [x]

tailend_reverse         lst   = tailend_reverse_aux [] lst
tailend_reverse_aux acc []    = acc
tailend_reverse_aux acc (x:t) = tailend_reverse_aux (x:acc) t

time :: IO t -> IO t
time fun = do
    start <- getCPUTime
    v <- fun
    end   <- getCPUTime
    let diff = (fromIntegral (end - start)) / (10^12)
    printf "%0.6f secs\n" (diff::Double)
    return v

bench fun funName siz = do
    printf "%s with %d: " (funName::[Char]) (siz::Int)
    let list = [1..siz]
    time $ fun list `seq` return ()

dummy [] = []
dummy (h:t) = dummy t

main = do
    bench dummy           "dummy"           10000000
    bench normal_reverse  "normal reverse"  1000000
    bench tailend_reverse "tailend reverse" 1000000
    bench normal_reverse  "normal reverse"  10000000
    bench tailend_reverse "tailend reverse" 10000000

File: picat/benchmarks/code/reverse.py
import time
import sys
sys.setrecursionlimit(20000)

def normal_reverse_list (inList):
    if inList == []:
        return []
    else:
        return [inList[-1]] + normal_reverse_list(inList[:-1])

def normal_reverse_tuple(inTuple):
    if inTuple == ():
        return ()
    else:
        return (inTuple[-1],) + normal_reverse_tuple(inTuple[:-1])

def tailend_reverse_list (inList):
    return tailend_reverse_list_aux(inList, [])
def tailend_reverse_list_aux(inList, acc):
    if inList == []:
        return acc
    else:
        return tailend_reverse_list_aux(inList[1:], [inList[0]]+acc)

def tailend_reverse_tuple (inTuple):
    return tailend_reverse_tuple_aux (inTuple, [])
def tailend_reverse_tuple_aux(inTuple, acc):
    if inTuple == ():
        return acc
    else:
        return tailend_reverse_tuple_aux(inTuple[1:], [inTuple[0]]+acc)

class Timer(object):
    def __init__(self, verbose=False):
        self.verbose = verbose

    def __enter__(self):
        self.start = time.time()
        return self

    def __exit__(self, *args):
        self.end = time.time()
        self.secs = self.end - self.start
        self.msecs = self.secs * 1000  # millisecs
        if self.verbose:
            print ('elapsed time: %f ms' % self.msecs)

def dummy(data):
    for i in range(len(data)):
        None

def bench(msg, call, lst, printFlag):
    res = []
    with Timer() as t:
        res = call(lst)
    print("%s : %s s" % (msg, t.secs))
    if (printFlag):
        print(res[:10])

size = 9000+1
bench("dummy", dummy, list(range(1, size)), False)
bench("normal reverse list", normal_reverse_list, list(range(1, size)), False)
bench("tailend reverse tuple", normal_reverse_tuple, tuple(range(1, size)), False)
bench("normal reverse list", tailend_reverse_list, list(range(1, size)), False)
bench("tailend reverse tuple", tailend_reverse_tuple, tuple(range(1, size)), False)
